File: Leetcode_2807.java
// LeetCode provides the ListNode class.
class Solution {
    public ListNode insertGreatestCommonDivisors(ListNode head) {
        ListNode curr = head;

        while (curr != null && curr.next != null) {
            int g = gcd(curr.val, curr.next.val);

            ListNode node = new ListNode(g); // new node holds the gcd
            node.next = curr.next;           // link new node to the next original node
            curr.next = node;                // link current node to the new node

            curr = node.next;                // move to next original node (skip inserted)
        }
        return head;
    }

    private int gcd(int a, int b) {
        // Euclid's algorithm (iterative is also fine)
        if (b == 0) return a;
        return gcd(b, a % b);
    }
}

File: Leetcode_2807.c
#include <stdlib.h>

// LeetCode's definition (or similar)
struct ListNode {
    int val;
    struct ListNode *next;
};

static int gcd(int a, int b) {
    // Iterative Euclid to avoid recursion overhead
    while (b != 0) {
        int t = a % b;
        a = b;
        b = t;
    }
    return a;
}

struct ListNode* insertGreatestCommonDivisors(struct ListNode* head) {
    struct ListNode* curr = head;

    while (curr != NULL && curr->next != NULL) {
        int g = gcd(curr->val, curr->next->val);

        struct ListNode* node = (struct ListNode*)malloc(sizeof(struct ListNode));
        node->val = g;                  // store gcd
        node->next = curr->next;        // new node points to next original node
        curr->next = node;              // current points to new node (insertion done)

        curr = node->next;              // move to next original node
    }
    return head;
}
